osu.Framework/Threading/Scheduler.cs

```diff
@@ -100,20 +100,29 @@ public void UpdateClock(IClock newClock)
         /// Run any pending work tasks.
         /// </summary>
         /// <returns>The number of tasks that were run.</returns>
-        public virtual int Update()
+        public int Update()
         {
-            lock (queueLock)
+            bool hasTimedTasks = timedTasks.Count > 0;
+            bool hasPerUpdateTasks = perUpdateTasks.Count > 0;
+
+            if (hasTimedTasks || hasPerUpdateTasks) // avoid taking out a lock if there are no items.
             {
-                queueTimedTasks();
-                queuePerUpdateTasks();
+                lock (queueLock)
+                {
+                    if (hasTimedTasks) queueTimedTasks();
```

Comment:
Would assume the same concern about per-update tasks applies to timed tasks as well?

```suggestion
                    if (timedTasks.Count > 0) queueTimedTasks();
```

Would also inline the local since it's not correct to store it in a local in the first place.

Reply:
No, the concern is only for the second, since the first method potentially adds the second type.

Reply:
I've removed the inner checks since they don't bring much value to the table.